### DISCUSSION TITLE
Does firejail rely on the Filesystem Hierarchy Standard (FHS)?

### DISCUSSION THREAD

**Discussion by amano-kenji** [netblue30/firejail#6198] [Q&A]
There are some linux distributions like nix os and gnu guix system that deviate from file hierarchy stndard.

**Comment by rusty-snake** [comment]
There are two things we need to consider here, firejail and firejail-profiles.

For profiles the answer is simple: yes

For firejail (the binary) it is more complex. Some basic requirements like `procfs` being mounted at `/proc` will likely always being depended on. Path specific commands such as `private-etc` will of course also depend on these paths. Otherwise it does not have to depend on it but does more than necessary ATM.

---

General, as IDK the layout of nix: If you use a sandbox unfriendly filesystem layout, you have to decide between tight sandboxes and working features.

---

Note that FHS isn't the only thing. Besides it there are `hier(7)`, XDG specs and `file-hierarchy(7)`.

**Reply by amano-kenji** [reply]
What I mean is that libraries are not in (/usr)/lib(64) and executables are not in (/usr)/(s)bin. There may be more deviations.

**Reply by amano-kenji** [reply]
https://guix.gnu.org/en/blog/2023/the-filesystem-hierarchy-standard-comes-to-guix-containers/ says

> Guix does not conform to the [Filesystem Hierarchy Standard](https://refspecs.linuxfoundation.org/fhs.shtml) (FHS). In practical terms, this means there is no global /lib containing libraries, /bin containing binaries,¹ and so on.

> However, this also leads to a difference in how many pieces of software expect their world to look like, relying on finding a library in /lib or an external tool in /bin. When these are hard coded and not overcome with appropriate build options, we patch code to refer to absolute paths in the store, like /gnu/store/hrgqa7m498wfavq4awai3xz86ifkjxdr-grep-3.6/bin/grep, to keep everything consistently contained within the store.

**Reply by amano-kenji** [reply]
I just chatted in #guix chat room. What I discovered is that

* (/usr)/lib(64) and (/usr)/(s)bin exist in `/gnu/store/random-hash`. firejail has to support whitelisting or blacklisting `/gnu/store/random-hash`. Libraries and executables are spread out across multiple random hash directories. Each version of each package has its own hash directory in `/gnu/store`.
* The only thing under /usr is /usr/bin/env on guix.

Can firejail support `/gnu/store/*/path/to/bin-or-lib`? If it can, it can also support nix because guix shares the fundamental premises with nix.

A quick fix for now is to not touch `/gnu/store/*`.

I guess going through every random hash directory in /gnu/store is going to be inefficient.

**Reply by rusty-snake** [reply]
There are three things.
- You want to deny `/gnu/store/foobar`
- You want to allow `/gnu/store/foobar`
- You want commands like private-etc to work.

And there are profiles.

**Reply by amano-kenji** [reply]
There is no way to manually deny or allow `/gnu/store/*`. Perhaps, you can programmatically do so by passing command line arguments to firejail.

**Reply by amano-kenji** [reply]
One package resides in /gnu/store/some-hash which contains symlinks to other hash directories of /gnu/store.

One gnu store hash directory contains symlinks to dependencies in other hash directories.

Thus, if you whitelist /gnu/store/some-hash and allow symlinks to point to other hash directories, you can sandbox a gnu guix system.

I haven't learned nix, but I read that guix is fundamentally the same as nix.

This makes me think that perhaps, private-bin is obsolete on guix and nix because there is no shared bin directory. There is no shared usr.

I will come with more information when I actually use gnu guix system.

**Reply by ghost** [reply]
Well, I'm a NixOS user, guix and nix are quite similar since the former is a fork of the latter. 

On to the point, while binaries in nix do indeed reside in the /nix/store directory, they are symlinked into the /run/current-system/sw/bin directory, or /etc/profiles/per-user/<user>/bin/ if you use home-manager.

I'm currently using firejail on NixOS and I haven't had any visible issues. Nonetheless, I noticed that firefox is able to access the entire /nix/store. 
The home directory is properly protected, and the network confinement options work fine too.

**Reply by amano-kenji** [reply]
@Praetoriam That's what I expect from firejail on guix and nix by default. Functional, but private-xxx directives are useless.

**Reply by ghost** [reply]
@amano-kenji Well, I suppose Apparmor can come in handy in these environments.

**Reply by amano-kenji** [reply]
With apparmor, I have to whitelist /nix/store anyway...

**Reply by ghost** [reply]
Have you checked `pkgs.apparmorRulesFromClosure`?. This thread helped me out a bit https://discourse.nixos.org/t/what-is-pkgs-apparmorrulesfromclosure/39423/2. Alas, documentation is pretty scarce when it comes to AppArmor.

**Reply by amano-kenji** [reply]
Where is that documented?

**Reply by amano-kenji** [reply]
Can firejail do a similar thing for guix and nix?

**Reply by topimiettinen** [reply]
Probably with some effort. Deep support for Nix paths could also be used to select allowed components based on the package contents, so instead of listing file names etc. in `private-etc`, indirection with packages could be specified. So for example, instead of `private-etc firefox`, there could be something like `private-etc @nixpkg:firefox`. This would result in allowing something like `/nix/store/gsbl1123dy7dil3vayj0w9irn0f2d4rf-firefox-124.0.2/etc` (guess, I don't know Nix). The same idea could apply also to traditional distros but it's less useful since the paths won't change. Perhaps a basic profile could be synthesized for any package out there with just standard templates like `private-etc @pkg:xyz` + `private-bin @pkg:xyz` + `private-lib @pkg:xyz @deps:pkg:xyz`?

**Reply by amano-kenji** [reply]
For nix and guix, `private-etc`, `private-lib`, and other private directives don't make sense. Just whitelisting all necessary store directories is more than sufficient.

Only strictly necessary files are symlinked in store directories.
